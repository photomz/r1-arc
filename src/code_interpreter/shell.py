"""
Interactive interpreter for ARC DSL with function tracing
"""

import code
import sys
import importlib
import json
import numpy as np
import typer
import os
import re
from pathlib import Path
from typing import Optional, Dict, List, Any, Tuple
import inspect

# Import local modules
from src.dsl.arc_types import *
from src.dsl.constants import *
from src.dsl.dsl import *
import src.dsl.main as main
from src.dsl.tracer import DSLTracer

# Task data directory
DATA_DIR = Path("data")

app = typer.Typer()


def list_dsl_functions() -> None:
    """List all available DSL functions"""
    dsl_funcs = [
        name
        for name, obj in inspect.getmembers(sys.modules["dsl"])
        if inspect.isfunction(obj) and not name.startswith("_")
    ]
    print("\nAvailable DSL functions:")
    print("------------------------")
    for func in sorted(dsl_funcs):
        print(f"- {func}")


def man_dsl_function(func_name: str) -> None:
    """Get type definition of a DSL function"""
    try:
        with open("out/dsl.pyi", "r") as f:
            pyi_contents = f.read()

        # Find the function definition using regex
        pattern = rf"def {func_name}\((.*?)\) -> (.*?):"
        match = re.search(pattern, pyi_contents, re.DOTALL)

        if match:
            args, return_type = match.groups()
            # Find docstring
            docstring_pattern = rf"def {func_name}.*?\n\s+\"\"\"(.*?)\"\"\""
            docstring_match = re.search(docstring_pattern, pyi_contents, re.DOTALL)
            docstring = (
                docstring_match.group(1).strip()
                if docstring_match
                else "No description available"
            )

            print(f"\nFunction: {func_name}")
            print("------------------------")
            print(f"Type: ({args}) -> {return_type}")
            print(f"Description: {docstring}")
        else:
            print(f"Function '{func_name}' not found in DSL")
    except FileNotFoundError:
        print("Type definitions file (dsl.pyi) not found")
    except Exception as e:
        print(f"Error getting function definition: {e}")


def load(task_id: str, idx: int = 0) -> Tuple[np.ndarray, Optional[np.ndarray]]:
    """Load a task's input grid and optionally its output

    Args:
        task_id: The task ID to load
        idx: Index of the example to load (0-based)

    Returns:
        Tuple of (input_grid, output_grid)
        Output grid may be None if idx is for a test example
    """
    # Use main.py's data loading
    data = main.get_data(train=True)

    if task_id not in data["train"]:
        raise ValueError(f"Task {task_id} not found")

    # Count total examples (train + test)
    train_examples = data["train"][task_id]
    test_examples = data["test"][task_id]
    all_examples = train_examples + test_examples

    # Check if index is valid
    if idx < 0 or idx >= len(all_examples):
        raise ValueError(f"Example index {idx} out of range (0-{len(all_examples)-1})")

    # Get the example
    example = all_examples[idx]

    # Convert to tuple of tuples
    input_grid = tuple(tuple(row) for row in example["input"])

    # Output may not exist for test examples
    output_grid = None
    if "output" in example:
        output_grid = tuple(tuple(row) for row in example["output"])

    return input_grid, output_grid


class DSLInterpreter(code.InteractiveConsole):
    """Interactive interpreter with DSL function tracing"""

    def __init__(self, task_id: Optional[str] = None):
        """Initialize the interpreter

        Args:
            task_id: Optional task ID to load initially
        """
        # Create a clean namespace
        local_ns = {}

        # Initialize the interpreter
        super().__init__(local_ns)

        # Initialize the tracer
        self.tracer = DSLTracer()
        self.current_task_id = None

        # Import DSL modules for the interpreter and instrument them
        for module_name in ["dsl", "constants", "arc_types"]:
            module = sys.modules[module_name]
            self.instrument_module(module)

        # Add helper functions to namespace
        self.locals["load"] = load
        self.locals["list"] = list_dsl_functions
        self.locals["man"] = man_dsl_function

        # Load task if specified
        if task_id:
            try:
                input_grid, _ = load(task_id)
                self.locals["I"] = input_grid
                self.current_task_id = task_id
                print(f"Loaded task {task_id}")
                print(f"Input grid 'I' shape: {input_grid.shape}")
            except Exception as e:
                print(f"Error loading task: {e}")

    def instrument_module(self, module):
        """Instrument a module and add it to the interpreter namespace"""
        # First instrument it
        self.tracer.instrument(module.__name__)

        # Then add the instrumented versions to our namespace
        for name, obj in module.__dict__.items():
            if not name.startswith("_"):  # Skip private members
                self.locals[name] = obj

    def runsource(self, source, filename="<input>", symbol="single"):
        """Override to add tracing for all executed code"""
        # Detect solver function definitions to auto-instrument
        solver_match = re.match(r"def\s+solve_([a-zA-Z0-9_]+)\s*\(", source)
        if solver_match:
            # Auto-load the task if it exists
            task_id = solver_match.group(1)
            if not self.current_task_id or self.current_task_id != task_id:
                try:
                    input_grid, _ = load(task_id)
                    self.locals["I"] = input_grid
                    self.current_task_id = task_id
                    print(f"Loaded task {task_id}")
                    print(f"Input grid 'I' shape: {input_grid.shape}")
                except Exception:
                    # If task not found, continue anyway
                    pass

        # Clear trace for each new execution
        self.tracer.clear()

        # Enable tracing
        self.tracer.enabled = True

        try:
            # Try to execute the code
            code_obj = code.compile_command(source, filename, symbol)
            if code_obj is None:
                # Incomplete input
                return True

            # Execute the code
            self.runcode(code_obj)

            # Process the trace
            self._process_trace()

            return False
        except (OverflowError, SyntaxError, ValueError):
            # Show syntax errors
            self.showsyntaxerror(filename)
            return False
        except Exception:
            # Show other errors
            self.showtraceback()
            return False

    def _process_trace(self):
        """Process and display the trace information"""
        # Get function stats
        stats = self.tracer.get_function_stats()

        if stats:
            print("\n--- Function Call Statistics ---")
            for name, data in sorted(
                stats.items(), key=lambda x: x[1]["count"], reverse=True
            ):
                print(f"{name}: {data['count']} calls, {data['avg_time_ms']:.2f}ms avg")
                # if data["grid_changes"]:
                #     print(f"  Grid changes: {data['grid_changes']}")

            # # Print function calls with grid changes
            # grid_change_calls = [
            #     call
            #     for call in self.tracer.execution_log
            #     if call.grid_changes and call.grid_changes > 0
            # ]

            # if grid_change_calls:
            #     print("\n--- Grid Transformations ---")
            #     for i, call in enumerate(grid_change_calls[:5]):
            #         print(f"{call.fname}: {call.grid_changes} changes")
            #         # Print first 5 coordinate changes
            #         for j, (i, j, old_val, new_val) in enumerate(call.diff_coords[:5]):
            #             print(f"  ({i},{j}): {old_val} -> {new_val}")

            #         if len(call.diff_coords) > 5:
            #             print(f"  ... and {len(call.diff_coords) - 5} more changes")

            # Show total call count
            print(f"\nTotal function calls: {len(self.tracer.execution_log)}")


@app.command()
def run(task_id: Optional[str] = None):
    """Run the DSL interpreter with tracing"""
    interpreter = DSLInterpreter(task_id)

    # Create welcome message
    print("DSL Interpreter with Tracing")
    print("---------------------------")
    print("Type Python/DSL code to evaluate")
    print("Available commands:")
    print("  load(task_id, idx=0): Load a specific task and example")
    print("  list(): List all available DSL functions")
    print("  man(func_name): Get type definition and description of a DSL function")
    print("  exit(): Exit the interpreter")

    # Start the interpreter
    interpreter.interact(banner="")


if __name__ == "__main__":
    app()
