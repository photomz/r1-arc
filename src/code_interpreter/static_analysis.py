"""
Code analysis module for the Python execution server.

This module contains functions for analyzing Python code, including:
- Collecting static metrics
- Analyzing AST nodes
- Detecting DSL function calls
"""

import ast
import json
from typing import Set
import libcst as cst
from dataclasses import dataclass
from pydantic import BaseModel, Field
from src.utils.devtools import debug
from src.utils import ROOT

dsl_info = json.load((ROOT / "src/dsl/out/dsl_typedefs.json").open())
dsl_functions = dsl_info["functions"]


class StaticMetrics(BaseModel):
    """AST static analysis of generated code literal"""

    n_lines: int = 0
    n_branch: int = 0
    n_loop: int = 0
    n_returns: int = 0
    n_funcs: int = 0
    n_dsl: Set[str] = Field(default_factory=set)
    imports: Set[str] = Field(default_factory=set)
    syntax_ok: bool = False
    # No error (str, code), has trace, same len I, O, intermediates
    run_ok: bool = False
    format_ok: bool = False

    @staticmethod
    def from_literal(codestr: str) -> "StaticMetrics":
        metrics = StaticMetrics()
        metrics.n_lines = len(lines := codestr.strip().split("\n"))
        metrics.format_ok = validate_solve_function_cst(codestr)

        try:
            tree = ast.parse(codestr)
            metrics.syntax_ok = True

            for node in ast.walk(tree):
                metrics = add_node(node, metrics)

            print("Syntax ok")
        except SyntaxError as e:
            debug(f"AST parsing failed: {e}")
            metrics.syntax_ok = False

        return metrics


def add_node(
    node: ast.AST,
    metrics: StaticMetrics,
) -> StaticMetrics:
    # Count node types
    node_type = type(node).__name__

    # Count specific constructs
    match node:
        case ast.FunctionDef():
            metrics.n_funcs += 1
        case ast.If() | ast.Match() | ast.Try():
            metrics.n_branch += 1
        case ast.For() | ast.While():
            metrics.n_loop += 1
        case ast.Return():
            metrics.n_returns += 1
        case ast.Import():
            for name in node.names:
                metrics.imports.add(name.name)
        case ast.ImportFrom() if node.module:
            metrics.imports.add(node.module)
        case ast.Call():
            if (
                isinstance(node.func, ast.Name)
                and (fid := node.func.id) in dsl_functions
            ):
                metrics.n_dsl.add(fid)
        case _:
            pass
    return metrics


class LoggingInjector(cst.CSTTransformer):
    """Injects logging code into a function definition."""

    def leave_FunctionDef(
        self, original_node: cst.FunctionDef, updated_node: cst.FunctionDef
    ) -> cst.FunctionDef:
        # Parse decorator from string
        decorator_module = cst.parse_module(
            "@log_sparse(stack_depth=10)\ndef dummy(): pass"
        )
        decorator = decorator_module.body[0].decorators[0]

        # Parse logging statement from string
        logging_code = cst.parse_statement(
            """
tracer.log_locals(list(locals().items()))"""
        )
        #         logging_code = cst.parse_statement(
        #             """
        # for name, value in locals().items():
        #     tracer.log_var(name, value)"""
        #         )

        # Get the function body statements
        body_statements = list(updated_node.body.body)

        # Check for return statements in SimpleStatementLine nodes
        has_return = False
        new_body = []

        for stmt in body_statements:
            if isinstance(stmt, cst.SimpleStatementLine) and any(
                isinstance(expr, cst.Return) for expr in stmt.body
            ):
                has_return = True
                new_body.append(logging_code)
            new_body.append(stmt)

        # If no return statements found, add logging at the end
        if not has_return:
            new_body.append(logging_code)

        # Update function with decorator and new body
        return updated_node.with_changes(
            # decorators=[decorator, *updated_node.decorators],
            body=updated_node.body.with_changes(body=new_body),
        )


def inject_logging(code0: str) -> str:
    # try:
    # Parse and transform the code
    module = cst.parse_module(code0)
    edited_tree = module.visit(LoggingInjector())

    # Verify the transformed code is valid Python
    code1 = edited_tree.code
    compile(code1, "<string>", "exec")
    return code1

    # except Exception as e:
    #     debug(str(e))
    #     return code0


@dataclass
class SolveFunctionValidator(cst.CSTVisitor):
    """A valid solver is of type solve: I -> O, and not redefined."""

    def __init__(self):
        super().__init__()
        self.has_solve_function = False
        self.n_param = 0
        self.n_return = 0
        self.current_function = None
        self.indent_depth = 0
        self.n_solvers = 0

    def visit_FunctionDef(self, node: cst.FunctionDef) -> None:
        self.indent_depth += 1

        if node.name.value == "solve" and self.indent_depth == 1:
            self.has_solve_function = True
            self.n_solvers += 1
            self.current_function = node.name.value
            self.n_param = len(node.params.params)

    def leave_FunctionDef(self, node: cst.FunctionDef) -> None:
        if node.name.value == "solve" and self.indent_depth == 1:
            self.current_function = None
        self.indent_depth -= 1

    def visit_Return(self, node: cst.Return) -> None:
        if self.current_function and self.current_function == "solve":
            self.n_return += 1

    @property
    def is_valid(self) -> bool:
        return (
            self.has_solve_function
            and self.n_param == 1
            and self.n_return == 1
            and self.n_solvers == 1
        )


def validate_solve_function_cst(code_str: str) -> bool:
    """Validates solve function using libcst."""
    try:
        module = cst.parse_module(code_str)
        validator = SolveFunctionValidator()
        module.visit(validator)
        return validator.is_valid
    except Exception as e:
        debug(e)
        return False


if __name__ == "__main__":
    code = """@
def solve_e9afcf9a(I):
    x1 = astuple(TWO, ONE)
    x2 = crop(I, ORIGIN, x1)
    x3 = hmirror(x2)
    x4 = hconcat(x2, x3)
    x5 = hconcat(x4, x4)
    O = hconcat(x5, x4)
    return O
    """.strip()

    debug(inject_logging(code))
