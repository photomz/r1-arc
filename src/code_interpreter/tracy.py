from src.dsl.arc_types import *
from src.dsl.constants import *
from src.dsl.dsl import *
from src.dsl.main import get_data

# from viztracer import log_sparse, VizTracer
import json
import tempfile
import sys


class VarTracer:  # (VizTracer):
    def __init__(self, *args, **kwargs):
        # super().__init__(
        #     *args, log_func_args=True, log_func_retval=True, log_print=True, **kwargs
        # )
        self.vartraces = []

    def log_locals(self, locale):
        # filter out non-serializable objects
        localvars = {}
        for name, value in locale:
            # if (
            #     not callable(value)
            #     and type(value) != "function"
            #     and isinstance(value, (int, float, str, tuple, list, dict))
            # ):
            # assert name in ["I", "O"] or name[0] == "x", name
            localvars[name] = value
        self.vartraces.append(localvars)

    def export(self):
        self.stop()

        return self.vartraces

        # Create a temporary file that gets automatically cleaned up
        with tempfile.NamedTemporaryFile(suffix=".json") as tmpf:
            self.save(tmpf.name)
            with open(tmpf.name, "r") as f:
                trace_data = json.load(f)
        var_events = list(
            filter(
                lambda d: d.get("cat") == "INSTANT",
                trace_data["traceEvents"],
            )
        )
        varlocals = {}
        for e in var_events:
            varliteral = e["args"]["object"]
            try:
                varvalue = eval(varliteral)  # WARN: Distorts 2d tuple shapes
                varlocals[e["name"]] = varvalue
            except SyntaxError as err:
                debug(e["name"], varliteral)
        self.clear()
        self.traces.append(varlocals)
        return varlocals


if __name__ == "__main__":
    key = "e9afcf9a"
    ex = get_data()["train"][key][0]

    I = ex["input"]

    tracer = VarTracer()
    tracer.start()

    @log_sparse(stack_depth=10)
    def solve_e9afcf9a(I):
        x1 = astuple(TWO, ONE)
        x2 = crop(I, ORIGIN, x1)
        x3 = hmirror(x2)
        x4 = hconcat(x2, x3)
        x5 = hconcat(x4, x4)
        O = hconcat(x5, x4)

        for name, value in locals().items():
            # can't move locals() to func, changes frame.
            tracer.log_var(name, value)
        return O

    O = solve_e9afcf9a(I)
    assert O == ex["output"]

    debug(tracer.export())
