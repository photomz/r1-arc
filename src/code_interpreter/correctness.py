from src.code_interpreter.execution import InterpreterOutput
from src.utils.tasks import deep_tuple
from dataclasses import dataclass


@dataclass
class InterpreterValidation:
    pass


# def validate(res: InterpreterOutput) -> InterpreterValidation:
#     assert trace.metrics.syntax_ok and trace.metrics.run_ok

#     shell_Is = [deep_tuple(run["I"]) for run in trace.traces]
#     debug(shell_Is)
#     assert shell_Is == test_Is

#     shell_Os = [deep_tuple(run["O"]) for run in trace.traces]
#     test_Os = [deep_tuple(ex.O) for ex in task.train + task.test]
#     debug(shell_Os, test_Os)
#     assert shell_Os == test_Os

#     assert trace == snapshot
