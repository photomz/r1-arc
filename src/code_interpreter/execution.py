"""
Minimal execution module for Python code in a safe environment.
"""

import json
import os
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, List, Dict

import jinja2
import subprocess

from src.dsl import *
from src.utils.devtools import debug
from src.utils.tasks import deep_tuple
from src.code_interpreter.static_analysis import StaticMetrics, inject_logging
from pygments import highlight
from pygments.formatters import TerminalFormatter
from pygments.lexers import PythonLexer
from src.utils import ROOT

# Setup Jinja2 environment for templates
TEMPLATE_DIR = Path(__file__).parent / "templates"
jinja_env = jinja2.Environment(
    loader=jinja2.FileSystemLoader(TEMPLATE_DIR),
    trim_blocks=True,
    lstrip_blocks=True,
)


ALLOWED_MODULES = {}  # TODO: Unused feature.
TIMEOUT = 10

TRACER_HOOK = "src.dsl.dsl"
INPUT_VAR = "I"
OUTPUT_VAR = "O"

pprint = lambda t: print(highlight(t, PythonLexer(), TerminalFormatter()))


@dataclass
class SubprocessResponse:
    stdout = ""
    stderr = ""
    return_code = 0
    timed_out = False


@dataclass
class TracerOutput:
    I: List[Grid]
    O: List[Grid]
    O_shapes: List[Shape]
    I_shapes: List[Shape]
    intermediates: List[Dict[str, Any]]

    @staticmethod
    def from_json(data: dict) -> "TracerOutput":
        return TracerOutput(
            I=[deep_tuple(i) for i in data["I"]],
            O=[deep_tuple(o) for o in data["O"]],
            O_shapes=[deep_tuple(shape) for shape in data["O_shapes"]],
            I_shapes=[deep_tuple(shape) for shape in data["I_shapes"]],
            intermediates=[
                {k: deep_tuple(v) for (k, v) in d.items()}
                for d in data["intermediates"]
            ],
        )


@dataclass
class InterpreterOutput:
    response: SubprocessResponse
    traces: TracerOutput
    codestring: str
    inputs: Any
    run_ok: bool = False


def render_template(code: str, input_data: Any) -> str:
    """Create templated code to execute in subprocess"""
    try:
        template = jinja_env.get_template("wrapper.jinja2")
        return template.render(
            code_literal=code,
            input=repr(input_data),
        )
    except jinja2.exceptions.TemplateError as e:
        debug(f"Error rendering template: {e}")
        raise RuntimeError(f"Template rendering failed: {e}")


def execute_in_subprocess(tempfp: str) -> SubprocessResponse:
    """Run a Python subprocess with the given file"""
    process = subprocess.Popen(
        [sys.executable, "-E", tempfp],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        env=os.environ.copy(),
        shell=False,
        cwd=os.path.dirname(tempfp),
    )

    res = SubprocessResponse()

    try:
        stdout, stderr = process.communicate(timeout=TIMEOUT)
        res.stdout = stdout
        res.stderr = stderr
        res.return_code = process.returncode
    except subprocess.TimeoutExpired:
        process.kill()
        stdout, stderr = process.communicate()
        res.stdout = stdout
        res.stderr = f"Code Interpreter timed out after {TIMEOUT}s"
        res.return_code = -1
        res.timed_out = True
    return res


@dataclass
class Interpreter:

    @classmethod
    def run(
        cls, codestring: str, inputs, id, raise_error=True, cleanup=True
    ) -> InterpreterOutput:
        """Execute Python code with transform function and return results"""

        # Create wrapped code and save to file
        injected_codestr = inject_logging(codestring)
        wrapped_code = render_template(injected_codestr, inputs)
        filename = f'{id}-{time.strftime("%Y%m%d_%H%M")}.py'
        tmp_dir = ROOT / "src/tmp"
        tmp_dir.mkdir(parents=True, exist_ok=True)
        file_path = tmp_dir / filename

        with open(file_path, "w") as f:
            f.write(wrapped_code)

        # Execute the code in a subprocess
        res = execute_in_subprocess(file_path)
        traces = None
        if tracefp := res.stdout.strip().split("\n")[-1]:  # See jinja2 template
            try:
                with open(tracefp, "r") as f:
                    traces = TracerOutput.from_json(json.load(f))
                    # debug(traces)
            except FileNotFoundError:
                debug(res.stdout)
                debug(tracefp)
        if res.stderr:
            debug(res.stdout)
            pprint(f"Subprocess {id} " + res.stderr)
            if raise_error:
                raise RuntimeError(res.stderr)

        if not traces:
            debug(injected_codestr)

        run_ok = (
            not res.stderr  # no error str
            and res.return_code == 0  # no error code
            and traces  # read trace
            and traces.O  # trace has output
            and traces.I  # trace has input
            and len(traces.O)
            == len(traces.I)
            == len(inputs)
            == len(traces.intermediates)  # io len match
        )

        if run_ok and cleanup:
            os.unlink(file_path)
            os.unlink(tracefp)

        return InterpreterOutput(
            response=res,
            traces=traces,
            run_ok=run_ok,
            codestring=codestring,
            inputs=inputs,
        )


# async def run_python_transform_async(*args, **kwargs):
#     return await asyncio.to_thread(run_python_transform_sync, *args, **kwargs)
