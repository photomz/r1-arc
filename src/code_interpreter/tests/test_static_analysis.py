from src.code_interpreter.static_analysis import (
    StaticMetrics,
    inject_logging,
    validate_solve_function_cst,
    SolveFunctionValidator,
)
import pytest

code = """
def foo():
    pass
    
def bar():
    return foo
    
def foobar():
    while _ in range(10):
        if _ % 4:
            return _
    return 0 # Dead

def solve_e9afcf9a(I):
    x1 = astuple(TWO, ONE)
    x2 = crop(I, ORIGIN, x1)
    x3 = hmirror(x2)
    x4 = hconcat(x2, x3)
    x5 = hconcat(x4, x4)
    O = hconcat(x5, x4)
    return O
    """.strip()

f_67a3c6ac = """
import os
def solve(I):
    O = vmirror(I)
    return O""".strip()

STRINGS = [code, f_67a3c6ac]


def test_ast(snapshot):
    """Should meter syntax by AST"""
    m = StaticMetrics.from_literal(code)
    assert m.n_funcs == 4
    assert m.n_returns == 4
    assert m == snapshot


@pytest.mark.parametrize("code", STRINGS)
def test_cst(snapshot, code):
    """Should inject logging code"""
    return inject_logging(code) == snapshot


def test_no_solver():
    """Should reject code without a solve function"""
    code = """
def not_solve(x):
    return x
    """
    valid = validate_solve_function_cst(code)
    assert not valid
    # assert isinstance(validator, SolveFunctionValidator)
    # assert not validator.has_solve_function
    # assert not validator.is_valid


def test_nested_solve_function():
    """Should reject nested solve functions that aren't at the top level"""
    code = """
def outer():
    def solve(x):
        return x
    return solve(5)
    """
    valid = validate_solve_function_cst(code)
    assert not valid
    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # assert not validator.is_valid
    # assert not validator.has_solve_function


def test_multiple_solve_functions():
    """Should detect multiple solve functions in the same module"""
    code = """
def solve(x):
    return x
def solve(y):
    return y*2
    """
    valid = validate_solve_function_cst(code)
    assert not valid
    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # # The validator will only track the last one it sees
    # assert validator.has_solve_function
    # assert not validator.is_valid
    # assert validator.n_return == 2


def test_conditional_returns():
    """Should reject solve functions with multiple return statements in conditionals"""
    code = """
def solve(x):
    if x > 0:
        return x
    else:
        return -x
    """
    valid = validate_solve_function_cst(code)
    assert not valid
    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # assert validator.has_solve_function
    # assert validator.n_param == 1
    # assert validator.n_return == 2
    # assert not validator.is_valid


def test_loop_return():
    """Should reject solve functions with returns inside loops"""
    code = """
def solve(x):
    for i in range(10):
        if i == x:
            return i
    return 0
    """
    valid = validate_solve_function_cst(code)
    assert not valid
    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # assert validator.has_solve_function
    # assert validator.n_param == 1
    # assert validator.n_return == 2
    # assert not validator.is_valid


def test_code_injection():
    """Should validate solve functions even with potentially malicious code"""
    code = """
def solve(x):
    exec("import os; os.system('rm -rf /')")
    return x
    """
    valid = validate_solve_function_cst(code)
    assert valid
    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # assert validator.has_solve_function
    # assert validator.n_param == 1
    # assert validator.n_return == 1
    # assert validator.is_valid  # Note: our validator only checks structure, not security


def test_extremely_large_function():
    """Should handle extremely large solve functions without crashing"""
    code = "def solve(x):\n" + "    y = x\n" * 10000 + "    return y"
    valid = validate_solve_function_cst(code)
    assert valid

    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # assert validator.has_solve_function
    # assert validator.n_param == 1
    # assert validator.n_return == 1
    # assert validator.is_valid


def test_unicode_deception():
    """Should reject functions with names that look like 'solve' but use different Unicode characters"""
    code = """
def sοlve(x):  # Uses Greek omicron (ο) instead of 'o'
    return x
    """
    assert not validate_solve_function_cst(code)
    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # assert not validator.has_solve_function
    # assert not validator.is_valid


def test_default_parameters():
    """Should handle solve functions with default parameters"""
    code = """
def solve(x=10):
    return x
    """
    assert (valid := validate_solve_function_cst(code))
    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # assert validator.has_solve_function
    # assert validator.n_param == 1
    # assert validator.n_return == 1
    # assert validator.is_valid


def test_return_with_side_effects():
    """Should validate solve functions with side effects in return statements"""
    code = """
def solve(x):
    return print(x) or x
    """
    assert (valid := validate_solve_function_cst(code))
    # validator = validate_solve_function_cst(code)
    # assert isinstance(validator, SolveFunctionValidator)
    # assert validator.has_solve_function
    # assert validator.n_param == 1
    # assert validator.n_return == 1
    # assert validator.is_valid


def test_syntax_error():
    """Should handle code with syntax errors gracefully"""
    code = """
def solve(x):
    return x)
    """
    valid = validate_solve_function_cst(code)
    assert not valid
    # validator = validate_solve_function_cst(code)
    # assert not validator  # Should return False for syntax errors


# def test_injection_security():
#     """Should safely handle injection attempts in logging transformation"""
#     code = """def solve(x):\n    return "\\"\nexec('import os; os.system(\\"rm -rf /\\")')\n#"""
#     result = inject_logging(code)
#     assert "exec(" not in result or "os.system" not in result

qwen_repr = """
def solve(I):
    # Find all 3s in the grid
    threes = ofcolor(I, 3)
    # For each 3, check if there's another 3 two cells away in any cardinal direction
    replacements = []
    for loc in threes:
        for direction in [(-2, 0), (2, 0), (0, -2), (0, 2)]:
            neighbor = (loc[0] + direction[0], loc[1] + direction[1])
            if neighbor in threes:
                replacements.append(loc)
                break
    # Create a new grid where the replacements are changed to 4
    O = canvas(0, shape(I))
    for loc in replacements:
        O = fill(O, 4, (loc,))
    return O
"""


def test_qwen_regression(snapshot):
    """Should analyse base Qwen 14B gen code"""
    assert StaticMetrics.from_literal(qwen_repr) == snapshot
