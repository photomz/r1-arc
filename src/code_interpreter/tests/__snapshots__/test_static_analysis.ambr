# serializer version: 1
# name: test_ast
  StaticMetrics(n_lines=20, n_branch=1, n_loop=1, n_returns=4, n_funcs=4, dsl_primitives={'astuple', 'hmirror', 'crop', 'hconcat'}, imports=set(), syntax_ok=True, run_ok=False, format_ok=False)
# ---
# name: test_cst[def foo():\n    pass\n    \ndef bar():\n    return foo\n    \ndef foobar():\n    while _ in range(10):\n        if _ % 4:\n            return _\n    return 0 # Dead\n\ndef solve_e9afcf9a(I):\n    x1 = astuple(TWO, ONE)\n    x2 = crop(I, ORIGIN, x1)\n    x3 = hmirror(x2)\n    x4 = hconcat(x2, x3)\n    x5 = hconcat(x4, x4)\n    O = hconcat(x5, x4)\n    return O]
  '''
  @log_sparse(stack_depth=10)
  def foo():
      pass
      
      tracer.log_locals(list(locals().items()))
      
  @log_sparse(stack_depth=10)
  def bar():
      
      tracer.log_locals(list(locals().items()))
      return foo
      
  @log_sparse(stack_depth=10)
  def foobar():
      while _ in range(10):
          if _ % 4:
              return _
      
      tracer.log_locals(list(locals().items()))
      return 0 # Dead
  
  @log_sparse(stack_depth=10)
  def solve_e9afcf9a(I):
      x1 = astuple(TWO, ONE)
      x2 = crop(I, ORIGIN, x1)
      x3 = hmirror(x2)
      x4 = hconcat(x2, x3)
      x5 = hconcat(x4, x4)
      O = hconcat(x5, x4)
      
      tracer.log_locals(list(locals().items()))
      return O
  '''
# ---
# name: test_cst[def foo():\n    pass\n    \ndef bar():\n    return foo\n\ndef solve_e9afcf9a(I):\n    x1 = astuple(TWO, ONE)\n    x2 = crop(I, ORIGIN, x1)\n    x3 = hmirror(x2)\n    x4 = hconcat(x2, x3)\n    x5 = hconcat(x4, x4)\n    O = hconcat(x5, x4)\n    return O]
  '''
  @log_sparse(stack_depth=10)
  def foo():
      pass
      
      tracer.log_locals(list(locals().items()))
      
  @log_sparse(stack_depth=10)
  def bar():
      
      tracer.log_locals(list(locals().items()))
      return foo
  
  @log_sparse(stack_depth=10)
  def solve_e9afcf9a(I):
      x1 = astuple(TWO, ONE)
      x2 = crop(I, ORIGIN, x1)
      x3 = hmirror(x2)
      x4 = hconcat(x2, x3)
      x5 = hconcat(x4, x4)
      O = hconcat(x5, x4)
      
      tracer.log_locals(list(locals().items()))
      return O
  '''
# ---
# name: test_cst[def solve_e9afcf9a(I):\n    x1 = astuple(TWO, ONE)\n    x2 = crop(I, ORIGIN, x1)\n    x3 = hmirror(x2)\n    x4 = hconcat(x2, x3)\n    x5 = hconcat(x4, x4)\n    O = hconcat(x5, x4)\n    return O]
  '''
  @log_sparse(stack_depth=10)
  def solve_e9afcf9a(I):
      x1 = astuple(TWO, ONE)
      x2 = crop(I, ORIGIN, x1)
      x3 = hmirror(x2)
      x4 = hconcat(x2, x3)
      x5 = hconcat(x4, x4)
      O = hconcat(x5, x4)
      
      tracer.log_locals(list(locals().items()))
      return O
  '''
# ---
# name: test_cst[import os\ndef solve(I):\n    O = vmirror(I)\n    return O]
  '''
  import os
  @log_sparse(stack_depth=10)
  def solve(I):
      O = vmirror(I)
      
      tracer.log_locals(list(locals().items()))
      return O
  '''
# ---
# name: test_qwen_regression
  StaticMetrics(n_lines=16, n_branch=1, n_loop=3, n_returns=1, n_funcs=1, dsl_primitives={'ofcolor', 'shape', 'fill', 'canvas'}, imports=set(), syntax_ok=True, run_ok=False, format_ok=True)
# ---
